### PR TITLE
feat(comm): add Blackwell MNNVL CuTe DSL all-reduce fusion backend

### PR DESCRIPTION
**Pull request by qiangyicheng**
## 📌 Description

This PR adds a BF16 Blackwell backend for the latency-critical tail of tensor-parallel MoE layers. It supports two patterns through the unified `allreduce_fusion` API:

- AllReduce + residual add + RMSNorm
- MoE finalize + shared-expert add + AllReduce + residual add + RMSNorm

The initial profiles target GB300, `H=8192`, `Top-K=10`, and TP8/TP16, covering decode through prefill. The implementation uses CuTe DSL, PyTorch Symmetric Memory, and MNNVL/NVLink multicast.

### Design goal

The goal is one backend that remains efficient from small-`M` decode to large-`M` prefill. For small payloads, LL and BT aggressively use Programmatic Dependent Launch (PDL): stages are enqueued together but keep independent grid, CTA, cluster, and warp configurations. At large `M`, HT switches to persistent warp-specialized execution.

| Protocol | Target region | Structure | Main idea |
|---|---|---|---|
| LL (low latency) | Small `M` / decode | 2 PDL-chained kernels | Publish to a symmetric mailbox, then run Lamport reduction + residual + RMSNorm. |
| BT (balanced) | Medium `M` | 3 PDL-chained kernels | Owner scatter, owner reduction/multicast, then materialization + RMSNorm. |
| HT (high throughput) | Large `M` / prefill | Persistent kernel | Warp-specialized finalize, communication, and RMSNorm pipeline. |

#### LL protocol

ll_protocol_two_kernel

Kernel 1 performs local finalize and publishes BF16 contributions with STMC into a triple-buffered symmetric mailbox. Kernel 2 performs the ordered Lamport reduction, residual add, optional prenorm output, and RMSNorm.

#### BT protocol

bt_protocol_three_kernel

Kernel 1 finalizes and unicasts shards to their owners. Kernel 2 reduces owner-local contributions, adds the residual, and broadcasts prenorm through STMC. Kernel 3 materializes the output and runs RMSNorm with its own launch configuration.

#### HT protocol

ht_protocol_warp_specialization

HT uses persistent CTAs with loader, finalize/RMS, publisher, and reduction warp roles. Contributions are reduced with LDMC over `H / TP` shards and prenorm values are published with STMC, overlapping communication and computation at large `M`.

### Dispatch and workspace integration

`MNNVLCuteDSLAllReduceFusionWorkspace` owns the compiled variants, symmetric state, and static routing profile. It compiles only the protocols required by the requested capacity and completes symmetric-memory rendezvous before use.

The default GB300 profile uses the following protocol boundaries:

| Path | TP | LL | BT | HT |
|---|---:|---:|---:|---:|
| MoE finalize + AR + residual + RMSNorm | 8 | `M <= 23` | `24–48` (P0), `49–703` (P1) | `M >= 704` |
| AR + residual + RMSNorm | 8 | `M <= 15` | `16–256` (P0), `257–1024` (P1) | `M >= 1025` |
| MoE finalize + AR + residual + RMSNorm | 16 | `M <= 7` | `8–52` (P0), `53–703` (P1) | `M >= 704` |
| AR + residual + RMSNorm | 16 | `M <= 5` | `6–512` (P0), `513–959` (P1) | `M >= 960` |

Profiles are keyed by TP size, hidden size, Top-K, and dtype, and map contiguous `M` ranges to a protocol and tuning preset. Runtime arguments are validated against the compiled workspace contract.

### Performance

gb300_fi_finalize_tp8_tp16_4panel_loglog

The figure reports maximum-rank latency on GB300 for BF16, `H=8192`, `Top-K=10`, TP8/TP16, and `M=1..8192`. The upper row includes MoE finalize; the lower row starts from a materialized rank-local contribution. Baselines use either FlashInfer MNNVL AR/residual/RMSNorm or NCCL AllReduce + fused add/RMSNorm. Every plotted point passed its reference check.

At the common power-of-two points covered by the default routed protocol, the current measurements provide the following conservative speedup ranges:

| Semantics | TP | vs. composed FlashInfer MNNVL baseline | vs. NCCL baseline |
|---|---:|---:|---:|
| With finalize | 8 | `1.74–3.20x` faster | `2.32–8.29x` faster |
| With finalize | 16 | `1.74–3.19x` faster | `2.48–11.36x` faster |
| Without finalize | 8 | `1.03–1.68x` faster | `1.62–7.93x` faster |
| Without finalize | 16 | `1.05–1.80x` faster | `1.81–11.30x` faster |

LL leads in decode, BT covers the middle range, and persistent HT scales best into prefill.

### Correctness strategy and numerical contract

Correctness uses two layers. The common end-to-end sweep checks the new backend and existing FlashInfer/NCCL paths against an independent unfused reference. Every one of the 70 graph invocations at each reported point is validated on every rank. Dedicated distributed tests then force LL, BT, and HT individually and apply a reference matching each protocol's reduction order.

| Protocol/path | Independent reference | Required prenorm result | Required RMSNorm result |
|---|---|---|---|
| LL | Gather every rank's BF16 contribution and accumulate in explicit rank order in FP32, then add residual and materialize BF16. | Bitwise equality for both all-reduce and finalize paths. | At most 1 BF16 ULP from RMSNorm recomputed from the produced prenorm, and at most 2 ULP end to end. |
| BT all-reduce | The same explicit rank-ordered FP32 reference; BT's owner decomposition is allowed its documented rounding point. | At most 1 BF16 ULP. | At most 1 ULP from the produced prenorm, and at most 2 ULP end to end. |
| BT finalize | Route-by-route FP32 local finalize followed by the explicit rank-ordered reference. | Bitwise equality. | At most 1 ULP from the produced prenorm, and at most 2 ULP end to end. |
| HT | PyTorch Symmetric Memory's NVLS `multimem_one_shot_all_reduce_out`, matching HT's multicast reduction semantics. | Bitwise equality for both all-reduce and finalize paths. | At most 1 ULP both from the produced prenorm and end to end. |

The dedicated tests also:

- Use order-sensitive `2^24`, `+1`, `-2^24` inputs to expose incorrect association order.
- Compute MoE finalize independently, route by route in FP32, including shared-expert addition.
- Check `residual_out` and `norm_out` separately, using raw BF16 bits or explicit ULP distance.
- Cover both fusion patterns, TP8/TP16, both BT presets, non-zero `weight_bias`, and every default routing boundary.

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used my preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see the [pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## Reviewer Notes

Please focus on the LL/BT/HT split and crossover points, PDL stage boundaries, symmetric-memory lifetime, and the protocol-specific bitwise/ULP contracts. The initial built-in profile is scoped to Blackwell BF16, GB300, `H=8192`, `Top-K=10`, and TP8/TP16.

The workspace must be fully constructed before its first invocation, and calls sharing one workspace must not overlap.

## Summary by CodeRabbit

* **New Features**
  * Added MNNVL CuTe DSL support for fused AllReduce and MoE finalization.
  * Added LL, BT, and HT protocol options with configurable presets and automatic workload-based selection.
  * Added residual handling, RMS normalization, shared-expert support, and BF16 execution.
  * Added symmetric-memory support for high-performance distributed operations.

* **Bug Fixes**
  * Improved workspace and execution validation, including tensor shapes, alignment, distributed setup, and supported environments.
  * Clarified tracing and error messages for supported MoE operations and workspace types.

### CODE REVIEW

**Review by jiahanc — APPROVED** [review]
LGTM

File: flashinfer/comm/mnnvl_cutedsl/kernel_bt/protocol.py [hunk 1]
```diff
@@ -0,0 +1,507 @@
...
+        gamma: torch.Tensor,
+        state: BTProtocolState,
+        norm_output: torch.Tensor,
+        residual_output: torch.Tensor | None,
+        m: int,
+    ) -> None:
+        residual_arg = residual_source if residual_source is not None else norm_output
+        residual_output_arg = (
+            residual_output if residual_output is not None else norm_output
+        )
```
[the remark this answers is not in the document]
Comment (qiangyicheng) [reply flashinfer/comm/mnnvl_cutedsl/kernel_bt/protocol.py:173]: Validation is performed by the public _mnnvl_cutedsl_allreduce_fusion entry point before these private protocol paths are invoked. The fallback tensors are used only for compile-time-disabled operands, and that contract is now documented on the workspace. We intentionally avoid duplicating public validation inside private protocol wrappers.

File: flashinfer/comm/mnnvl_cutedsl/kernel_bt/protocol.py [hunk 2]
```diff
@@ -0,0 +1,507 @@
...
+        }
+        device_kernel: Any
+        if tuning.elements_per_thread == 1:
+            device_kernel = _ScalarFinalizeUnicastDeviceKernel(**kwargs)
+        elif tuning.elements_per_thread == VEC_BF16:
+            device_kernel = _VectorFinalizeUnicastDeviceKernel(**kwargs)
+        else:
+            device_kernel = _NarrowVectorFinalizeUnicastDeviceKernel(
+                **kwargs, elements_per_thread=tuning.elements_per_thread
+            )
```
[the remark this answers is not in the document]
Comment (qiangyicheng) [reply flashinfer/comm/mnnvl_cutedsl/kernel_bt/protocol.py:380]: Addressed. elements_per_thread is validated in the BT compile path, while hidden-size vector alignment is enforced once by StaticProfile before any protocol is constructed.

File: flashinfer/comm/mnnvl_cutedsl/kernel_bt/protocol.py [hunk 3]
```diff
@@ -0,0 +1,507 @@
...
+            device,
+            group,
+            require_multicast=True,
+        )
+        prenorm.tensor.view(torch.int16).fill_(-32768)
+        return BTProtocolState(
+            contribution_mailbox=contribution,
+            prenorm_mailbox=prenorm,
+            stage_state=torch.zeros((2,), dtype=torch.uint32, device=device),
+        )
```
[the remark this answers is not in the document]
Comment (qiangyicheng) [reply flashinfer/comm/mnnvl_cutedsl/kernel_bt/protocol.py:511]: Addressed at the workspace boundary. The current device is synchronized and the process group is barriered once after all selected protocol states have been initialized, covering LL, BT, and HT without redundant per-protocol barriers.

File: flashinfer/comm/mnnvl_cutedsl/kernel_ll/protocol.py [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+            to_cute_dynamic(
+                residual_output_arg.flatten(),
+                16,
+                divisibility=self.hidden_size,
+            ),
+            to_cute_dynamic(norm_output.flatten(), 16, divisibility=self.hidden_size),
+            to_cute(state.stage_state, 4),
+            Int32(m),
+            current_cu_stream(),
+        )
```
[the remark this answers is not in the document]
Comment (qiangyicheng) [reply flashinfer/comm/mnnvl_cutedsl/kernel_ll/protocol.py:181]: The public fusion entry point already requires residual_in whenever the compiled workspace enables residual addition. These protocol wrappers are private implementation paths, so we keep them lightweight and rely on the validated workspace contract rather than duplicating the check.

File: flashinfer/comm/mnnvl_cutedsl/symmetric_buffer.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+            addresses = [
+                handle.get_remote_tensor(
+                    peer,
+                    tensor.shape,
+                    tensor.dtype,
+                ).data_ptr()
+                for peer in range(dist.get_world_size(group))
+            ]
+            if any(not address for address in addresses):
+                raise RuntimeError("Symmetric peer mapping is unavailable")
```
[the remark this answers is not in the document]
Comment (qiangyicheng) [reply flashinfer/comm/mnnvl_cutedsl/symmetric_buffer.py:89]: This backend follows the existing FlashInfer MNNVL integration pattern for PyTorch Symmetric Memory. Setting a repository-wide PyTorch version floor for these private APIs is a broader dependency-policy decision and is outside the scope of this backend PR.

**Review by aleozlx — APPROVED** [review]
lgtm

**Review by samuellees — APPROVED** [review]
LGTM
